File: packages/react-native-notificare-scannables/android/src/main/java/re/notifica/scannables/react_native/NotificareScannablesModule.kt
package re.notifica.scannables.react_native

import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.Promise
import re.notifica.Notificare
import re.notifica.NotificareCallback
import re.notifica.scannables.NotificareScannables
import re.notifica.scannables.ktx.scannables
import re.notifica.scannables.models.NotificareScannable

public class NotificareScannablesModule internal constructor(context: ReactApplicationContext) :
    NotificareScannablesModuleSpec(context), NotificareScannables.ScannableSessionListener {

    override fun initialize() {
        super.initialize()

        logger.hasDebugLoggingEnabled = Notificare.options?.debugLoggingEnabled ?: false

        EventBroker.setup(reactApplicationContext)
        Notificare.scannables().addListener(this)
    }

    override fun invalidate() {
        super.invalidate()

        Notificare.scannables().removeListener(this)
    }

    override fun getName(): String {
        return NAME
    }

    @ReactMethod
    override fun addListener(eventName: String) {
        // Keep: Required for RN built in Event Emitter Calls.
    }

    @ReactMethod
    override fun removeListeners(count: Double) {
        // Keep: Required for RN built in Event Emitter Calls.
    }

    // region NotificareScannables

    @ReactMethod
    override fun canStartNfcScannableSession(promise: Promise) {
        promise.resolve(Notificare.scannables().canStartNfcScannableSession)
    }

    @ReactMethod
    override fun startScannableSession(promise: Promise) {
        val activity = currentActivity ?: run {
            promise.reject(DEFAULT_ERROR_CODE, "Cannot start a scannable session without an activity.")
            return
        }

        Notificare.scannables().startScannableSession(activity)
        promise.resolve(null)
    }

    @ReactMethod
    override fun startNfcScannableSession(promise: Promise) {
        val activity = currentActivity ?: run {
            promise.reject(DEFAULT_ERROR_CODE, "Cannot start a scannable session without an activity.")
            return
        }

        Notificare.scannables().startNfcScannableSession(activity)
        promise.resolve(null)
    }

    @ReactMethod
    override fun startQrCodeScannableSession(promise: Promise) {
        val activity = currentActivity ?: run {
            promise.reject(DEFAULT_ERROR_CODE, "Cannot start a scannable session without an activity.")
            return
        }

        Notificare.scannables().startQrCodeScannableSession(activity)
        promise.resolve(null)
    }

    @ReactMethod
    override fun fetch(tag: String, promise: Promise) {
        Notificare.scannables().fetch(tag, object : NotificareCallback<NotificareScannable> {
            override fun onSuccess(result: NotificareScannable) {
                try {
                    promise.resolve(result.toJson().toReactMap())
                } catch (e: Exception) {
                    promise.reject(DEFAULT_ERROR_CODE, e)
                }
            }

            override fun onFailure(e: Exception) {
                promise.reject(DEFAULT_ERROR_CODE, e)
            }
        })
    }

    // endregion

    // region NotificareScannables.ScannableSessionListener

    override fun onScannableDetected(scannable: NotificareScannable) {
        try {
            EventBroker.dispatchEvent("re.notifica.scannables.scannable_detected", scannable.toJson().toReactMap())
        } catch (e: Exception) {
            logger.error("Failed to emit the re.notifica.scannables.scannable_detected event.", e)
        }
    }

    override fun onScannableSessionError(error: Exception) {
        try {
            EventBroker.dispatchEvent("re.notifica.scannables.scannable_session_failed", error.localizedMessage)
        } catch (e: Exception) {
            logger.error("Failed to emit the re.notifica.scannables.scannable_session_failed event.", e)
        }
    }

    // endregion

    public companion object {
        internal const val NAME = "NotificareScannablesModule"
        internal const val DEFAULT_ERROR_CODE = "notificare_error"
    }
}

File: packages/react-native-notificare-scannables/ios/NotificareScannablesModule.swift
import React
import NotificareKit
import NotificareScannablesKit
import UIKit

private let DEFAULT_ERROR_CODE = "notificare_error"

@objc(NotificareScannablesModule)
class NotificareScannablesModule: RCTEventEmitter {
    
    private var hasListeners = false
    private var eventQueue = [(name: String, payload: Any?)]()
    
    private var rootViewController: UIViewController? {
        get {
            UIApplication.shared.delegate?.window??.rootViewController
        }
    }
    
    override init() {
        super.init()
        
        Notificare.shared.scannables().delegate = self
    }
    
    override class func requiresMainQueueSetup() -> Bool {
        return true
    }
    
    override func startObserving() {
        hasListeners = true
        
        if !eventQueue.isEmpty {
            NotificareLogger.debug("Processing event queue with \(eventQueue.count) items.")
            eventQueue.forEach { sendEvent(withName: $0.name, body: $0.payload)}
            eventQueue.removeAll()
        }
    }
    
    override func stopObserving() {
        hasListeners = false
    }
    
    override func supportedEvents() -> [String] {
        return [
            "re.notifica.scannables.scannable_detected",
            "re.notifica.scannables.scannable_session_failed",
        ]
    }
    
    private func dispatchEvent(_ name: String, payload: Any?) {
        if hasListeners {
            sendEvent(withName: name, body: payload)
        } else {
            eventQueue.append((name: name, payload: payload))
        }
    }
    
    // MARK: - Notificare Scannables
    
    @objc
    func canStartNfcScannableSession(_ resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
        resolve(Notificare.shared.scannables().canStartNfcScannableSession)
    }
    
    @objc
    func startScannableSession(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        DispatchQueue.main.async {
            guard let rootViewController = self.rootViewController else {
                reject(DEFAULT_ERROR_CODE, "Cannot start a scannable session with a nil root view controller.", nil)
                return
            }
            
            Notificare.shared.scannables().startScannableSession(controller: rootViewController)
            resolve(nil)
        }
    }
    
    @objc
    func startNfcScannableSession(_ resolve: @escaping RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
        DispatchQueue.main.async {
            Notificare.shared.scannables().startNfcScannableSession()
            resolve(nil)
        }
    }
    
    @objc
    func startQrCodeScannableSession(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        DispatchQueue.main.async {
            guard let rootViewController = self.rootViewController else {
                reject(DEFAULT_ERROR_CODE, "Cannot start a scannable session with a nil root view controller.", nil)
                return
            }
            
            Notificare.shared.scannables().startQrCodeScannableSession(controller: rootViewController, modal: true)
            resolve(nil)
        }
    }
    
    @objc
    func fetch(_ tag: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        Notificare.shared.scannables().fetch(tag: tag) { result in
            switch result {
            case let .success(scannable):
                do {
                    let payload = try scannable.toJson()
                    resolve(payload)
                } catch {
                    reject(DEFAULT_ERROR_CODE, error.localizedDescription, nil)
                }
            case let .failure(error):
                reject(DEFAULT_ERROR_CODE, error.localizedDescription, nil)
            }
        }
    }
}

extension NotificareScannablesModule: NotificareScannablesDelegate {
    func notificare(_ notificareScannables: NotificareScannables, didDetectScannable scannable: NotificareScannable) {
        do {
            dispatchEvent("re.notifica.scannables.scannable_detected", payload: try scannable.toJson())
        } catch {
            NotificareLogger.error("Failed to emit the re.notifica.scannables.scannable_detected event.", error: error)
        }
    }
    
    func notificare(_ notificareScannables: NotificareScannables, didInvalidateScannerSession error: Error) {
        dispatchEvent("re.notifica.scannables.scannable_session_failed", payload: error.localizedDescription)
    }
}
